src/BatteryStateOfChargeService.cpp

```diff
@@ -16,21 +20,57 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    double diffOfTimeMSec = prevTime->msecsTo(myTime);
+    double diffOfTime = diffOfTimeMSec/3600000;
+    double avgCurrent = (myCurrent+prevCurrent)/2;
+
+    double changeInAmpHour = avgCurrent * diffOfTime;
+    ampHourTotal += changeInAmpHour;
+    //if(ampHourTotal > BATTERY_AMP_HOUR_CAPACITY){ampHourTotal = BATTERY_AMP_HOUR_CAPACITY;}
+    ampHourUsed = BATTERY_AMP_HOUR_CAPACITY- ampHourTotal;
+    return ampHourUsed;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if(myCurrent>=0)
+    {
+        return false;
+    }
+    return true;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    double ampHoursLeft;
+    if(isCharging())
+    {
+        ampHoursLeft = BATTERY_AMP_HOUR_CAPACITY - ampHourTotal;
+    }
+    else
+    {
+        ampHoursLeft = ampHourTotal;
+    }
+    int current = myCurrent;
+    if(myCurrent < 0)
+    {
+        current *= -1;
+    }
+    double msecLeft = (ampHoursLeft * 3600000) / current;
```

Comment:
This `3600000` is what we refer to as a "magic number" and should be abstracted into a constant in an anonymous namespace. We usually do that by adding to the top:
```
namespace 
{
    int CONVERSION_FACTOR = 3600000;
}
```
Of course, pick a better name than `CONVERSION_FACTOR` :)